MAJOR: Improved the Hostnet ruleset

ping-localhost opened the pull request:
This is a backwards incompatible change for the fact that all `phpcs` commands will fail once this has been released.

The tool itself won't pass PHPCS once this has been merged into master, but I'll create a separate PR for that as to not pollute the current one.

---
**New sniff**

I've added a new sniff (`Hostnet.Classes.CopyrightMustBeBeforeDeclare`) that forces our copyright docblock to be defined before `declare(strict_types=1)`. The sniff is able to automatically fix it for the cases I've seen.

---
**Downsides:**

A few of the checks cannot be resolved automatically, which is too bad, I'll fix those by hand (yes really):
- `Generic.PHP.BacktickOperator`
- `Generic.Files.LineLength`
- `Squiz.Operators.ValidLogicalOperators.NotAllowed`
- `SlevomatCodingStandard.Classes.UnusedPrivateElements.UnusedProperty`

There might be a few more, but it's doesn't happen that often and it's quite easy to resolve.

--- 
**Benchmark:** 

Running `phpcs` with the current ruleset on one of our internal applications takes `1 mins, 27.05 secs`. With the changed ruleset the total time for `phpcs` was `1 mins, 14.48 secs`, however `phpcbf` took `3 mins, 53.41 secs` .

_EDIT_: This benchmark was done before the sniff `Hostnet.Classes.CopyrightMustBeBeforeDeclare` was added.

ping-localhost committed:
MAJOR: Removed @author check since we no longer want to use it

ping-localhost committed:
MAJOR: Added extra sniffs to be more strict

ping-localhost committed:
We no longer support PHP7.0

ping-localhost committed:
Also run tests on PHP7.2

linaori commented on src/Hostnet/ruleset.xml:
```diff
@@ -1,73 +1,332 @@
+    <rule ref="Generic.CodeAnalysis.UnnecessaryFinalModifier"/>
+
+    <!-- Forbid useless empty method overrides -->
+    <rule ref="Generic.CodeAnalysis.UselessOverridingMethod"/>
```
Does this one take into account that the visibility _can_ change? I've seen this happen for a test once

linaori replied:
This should be allowed:

```php
class A
{
    private function foo()
    {

    }
}

class B extends A
{
    protected function foo()
    {

    }
}

class C extends B
{
    public function foo()
    {

    }
}
```

ping-localhost replied:
The sniff does not complain about this.

linaori commented on src/Hostnet/ruleset.xml:
```diff
@@ -1,73 +1,332 @@
+    <!-- Forbid dead code -->
+    <rule ref="SlevomatCodingStandard.Classes.UnusedPrivateElements">
+        <!-- Except private variables that are not used -->
+        <exclude name="SlevomatCodingStandard.Classes.UnusedPrivateElements.WriteOnlyProperty"/>
```
This is where you only write to them to keep a reference alive?

ping-localhost replied:
For example, yes.

linaori commented on src/Hostnet/ruleset.xml:
```diff
@@ -1,73 +1,332 @@
+    </rule>
+
+    <!-- Require comma after last element in multi-line array -->
+    <rule ref="SlevomatCodingStandard.Arrays.TrailingArrayComma"/>
```
THANK YOU!

ping-localhost replied:
I've just updated the ruleset to work with the code snippet you sent me.

ping-localhost committed:
Remove rules already covered in PSR2 and PSR1

linaori commented on src/Hostnet/ruleset.xml:
```diff
@@ -1,73 +1,332 @@
 
-    <exclude name="Hostnet.Commenting.FileCommentCopyright.MissingVersion"/>
+    <!-- Force `self::` for self-reference, force lower-case self, forbid spaces around `::` -->
+    <rule ref="Squiz.Classes.SelfMemberReference"/>
```
Be careful that this does not always work, how does this check treat scope changes in anonymous classes?

ping-localhost replied:
It takes anonymous classes into account as shown in it's tests:
- https://github.com/squizlabs/PHP_CodeSniffer/blob/master/src/Standards/Squiz/Tests/Classes/SelfMemberReferenceUnitTest.inc
- https://github.com/squizlabs/PHP_CodeSniffer/blob/master/src/Standards/Squiz/Tests/Classes/SelfMemberReferenceUnitTest.inc.fixed

linaori reviewed (commented):
Please double check some of my comments to see if they are not going to break anything

ping-localhost committed:
Ensured that we can declare arrays like we are used to

linaori reviewed (approved):
What about the readme?

ping-localhost committed:
Added sniff to ensure declare strict is after docblock

hboomsma commented:
* `SlevomatCodingStandard.Classes.UnusedPrivateElements.UnusedProperty` How does this behave with traits using the private properties (for example those generated by the `accessor-generator-plugin-lib`

ping-localhost committed:
Disable the UnusedProperty sniff

ping-localhost committed:
Improved the README.md

ping-localhost commented:
@iltar I've updated the [`README.md`](https://github.com/ping-localhost/phpcs-tool/blob/improve-ruleset/README.md) to be more slick and added a Wiki for all the rules we import and our custom sniffs.
@hboomsma It, indeed, doesn't work with the `accessor-generator-plugin-lib` so I've excluded it.

linaori reviewed (approved):
> It, indeed, doesn't work with the accessor-generator-plugin-lib so I've excluded it.

All the more reasons to not use traits :joy:
